File: src/main/java/io/github/subiyacryolite/jds/JdsDbOracle.java
package io.github.subiyacryolite.jds;

import com.javaworld.NamedPreparedStatement;
import io.github.subiyacryolite.jds.enums.JdsComponent;
import io.github.subiyacryolite.jds.enums.JdsComponentType;
import io.github.subiyacryolite.jds.enums.JdsImplementation;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

/**
 * Created by ifunga on 14/07/2017.
 */
public abstract class JdsDbOracle extends JdsDb {

    public JdsDbOracle() {
        supportsStatements = true;
        implementation = JdsImplementation.ORACLE;
    }

    @Override
    public int tableExists(Connection connection, String tableName) {
        int toReturn = 0;
        String sql = "SELECT COUNT(*) AS Result FROM all_objects WHERE object_type IN ('TABLE') AND object_name = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, tableName.toUpperCase());
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    toReturn = resultSet.getInt("Result");
                }
            }
        } catch (Exception ex) {
            toReturn = 0;
            ex.printStackTrace(System.err);
        }
        return toReturn;
    }

    @Override
    public int viewExists(Connection connection, String viewName) {
        int toReturn = 0;
        String sql = "SELECT COUNT(*) AS Result FROM all_objects WHERE object_type IN ('VIEW') AND object_name = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, viewName.toUpperCase());
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    toReturn = resultSet.getInt("Result");
                }
            }
        } catch (Exception ex) {
            toReturn = 0;
            ex.printStackTrace(System.err);
        }
        return toReturn;
    }

    @Override
    public int procedureExists(Connection connection, String procedureName) {
        int toReturn = 0;
        String sql = "SELECT COUNT(*) AS Result FROM all_objects WHERE object_type IN ('PROCEDURE') AND object_name = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, procedureName.toUpperCase());
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    toReturn = resultSet.getInt("Result");
                }
            }
        } catch (Exception ex) {
            toReturn = 0;
            ex.printStackTrace(System.err);
        }
        return toReturn;
    }

    @Override
    public int columnExists(Connection connection, String tableName, String columnName) {
        int toReturn = 0;
        String sql = "SELECT COUNT(COLUMN_NAME) AS Result FROM ALL_TAB_COLUMNS WHERE TABLE_NAME = :tableName AND COLUMN_NAME = :columnName";
        try (NamedPreparedStatement preparedStatement = new NamedPreparedStatement(connection, sql)) {
            preparedStatement.setString("tableName", tableName.toUpperCase());
            preparedStatement.setString("columnName", columnName.toUpperCase());
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    toReturn = resultSet.getInt("Result");
                }
            }
        } catch (Exception ex) {
            toReturn = 0;
            ex.printStackTrace(System.err);
        }
        return toReturn;
    }

    public String createOrAlterView(String viewName, String viewSql) {
        StringBuilder sb = new StringBuilder("CREATE VIEW\t");
        sb.append(viewName);
        sb.append("\tAS\t");
        sb.append(viewSql);
        String toExecute = sb.toString();
        return toExecute;
    }

    @Override
    protected void createStoreEntityInheritance(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreEntityInheritance.sql");
    }

    @Override
    protected void createStoreText(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreText.sql");
    }

    @Override
    protected void createStoreDateTime(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreDateTime.sql");
    }

    @Override
    protected void createStoreZonedDateTime(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreZonedDateTime.sql");
    }

    @Override
    protected void createStoreInteger(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreInteger.sql");
    }

    @Override
    protected void createStoreFloat(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreFloat.sql");
    }

    @Override
    protected void createStoreDouble(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreDouble.sql");
    }

    @Override
    protected void createStoreLong(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreLong.sql");
    }

    @Override
    protected void createStoreTextArray(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreTextArray.sql");
    }

    @Override
    protected void createStoreDateTimeArray(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreDateTimeArray.sql");
    }

    @Override
    protected void createStoreIntegerArray(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreIntegerArray.sql");
    }

    @Override
    protected void createStoreFloatArray(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreFloatArray.sql");
    }

    @Override
    protected void createStoreDoubleArray(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreDoubleArray.sql");
    }

    @Override
    protected void createStoreLongArray(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreLongArray.sql");
    }

    @Override
    protected void createStoreEntities(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createRefEntities.sql");
    }

    @Override
    protected void createRefEnumValues(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createRefEnumValues.sql");
    }

    @Override
    protected void createRefFields(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createRefFields.sql");
    }

    @Override
    protected void createRefFieldTypes(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createRefFieldTypes.sql");
    }

    @Override
    protected void createBindEntityFields(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createBindEntityFields.sql");
    }

    @Override
    protected void createBindEntityEnums(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createBindEntityEnums.sql");
    }

    @Override
    protected void createRefEntityOverview(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreEntityOverview.sql");
    }

    @Override
    protected void createRefOldFieldValues(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreOldFieldValues.sql");
        //allow multiple leaves you open to SLQ injection. Thus manually add these indexes here unless you want to add more files
        //oracle jdbc hates semi-colons
        executeSqlFromString(connection, "CREATE INDEX IntegerValues  ON JdsStoreOldFieldValues(EntityGuid, FieldId, Sequence, IntegerValue)");
        executeSqlFromString(connection, "CREATE INDEX FloatValues    ON JdsStoreOldFieldValues(EntityGuid, FieldId, Sequence, FloatValue)");
        executeSqlFromString(connection, "CREATE INDEX DoubleValues   ON JdsStoreOldFieldValues(EntityGuid, FieldId, Sequence, DoubleValue)");
        executeSqlFromString(connection, "CREATE INDEX LongValues     ON JdsStoreOldFieldValues(EntityGuid, FieldId, Sequence, LongValue)");
        executeSqlFromString(connection, "CREATE INDEX DateTimeValues ON JdsStoreOldFieldValues(EntityGuid, FieldId, Sequence, DateTimeValue)");
        executeSqlFromString(connection, "CREATE INDEX TextBlobValues ON JdsStoreOldFieldValues(EntityGuid, FieldId, Sequence)");
    }

    @Override
    protected void createStoreEntityBinding(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreEntityBinding.sql");
    }

    protected void createStoreTime(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreTime.sql");
    }

    protected void createStoreBlob(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createStoreBlob.sql");
    }

    @Override
    protected void createRefInheritance(Connection connection) {
        executeSqlFromFile(connection, "sql/oracle/createRefInheritance.sql");
    }

    @Override
    protected void prepareCustomDatabaseComponents(Connection connection) {
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_BLOB);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_TEXT);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_LONG);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_INTEGER);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_FLOAT);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_DOUBLE);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_DATE_TIME);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_TIME);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_ZONED_DATE_TIME);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_ENTITY_V_2);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_ENTITY_FIELDS);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_ENTITY_ENUMS);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_CLASS_NAME);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_ENUM_VALUES);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_FIELD_NAMES);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_FIELD_TYPES);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.MAP_ENTITY_INHERITANCE);
        prepareDatabaseComponent(connection, JdsComponentType.STORED_PROCEDURE, JdsComponent.SAVE_ENTITY_INHERITANCE);
    }

    @Override
    protected void prepareCustomDatabaseComponents(Connection connection, JdsComponent jdsComponent) {
        switch (jdsComponent) {
            case SAVE_ENTITY_V_2:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreEntityOverviewV2.sql");
                break;
            case SAVE_ENTITY_INHERITANCE:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreEntityInheritance.sql");
                break;
            case MAP_FIELD_NAMES:
                executeSqlFromFile(connection, "sql/oracle/procedures/procBindFieldNames.sql");
                break;
            case MAP_FIELD_TYPES:
                executeSqlFromFile(connection, "sql/oracle/procedures/procBindFieldTypes.sql");
                break;
            case SAVE_BLOB:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreBlob.sql");
                break;
            case SAVE_TIME:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreTime.sql");
                break;
            case SAVE_TEXT:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreText.sql");
                break;
            case SAVE_LONG:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreLong.sql");
                break;
            case SAVE_INTEGER:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreInteger.sql");
                break;
            case SAVE_FLOAT:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreFloat.sql");
                break;
            case SAVE_DOUBLE:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreDouble.sql");
                break;
            case SAVE_DATE_TIME:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreDateTime.sql");
                break;
            case SAVE_ZONED_DATE_TIME:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreZonedDateTime.sql");
                break;
            case SAVE_ENTITY:
                executeSqlFromFile(connection, "sql/oracle/procedures/procStoreEntityOverview.sql");
                break;
            case MAP_ENTITY_FIELDS:
                executeSqlFromFile(connection, "sql/oracle/procedures/procBindEntityFields.sql");
                break;
            case MAP_ENTITY_ENUMS:
                executeSqlFromFile(connection, "sql/oracle/procedures/procBindEntityEnums.sql");
                break;
            case MAP_CLASS_NAME:
                executeSqlFromFile(connection, "sql/oracle/procedures/procRefEntities.sql");
                break;
            case MAP_ENUM_VALUES:
                executeSqlFromFile(connection, "sql/oracle/procedures/procRefEnumValues.sql");
                break;
            case MAP_ENTITY_INHERITANCE:
                executeSqlFromFile(connection, "sql/oracle/procedures/procBindParentToChild.sql");
                break;
        }
    }
}

File: src/main/kotlin/io/github/subiyacryolite/jds/JdsDbOracle.kt
/**
 * Jenesis Data Store Copyright (c) 2017 Ifunga Ndana. All rights reserved.
 *
 * 1. Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 2. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 3. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * Neither the name Jenesis Data Store nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package io.github.subiyacryolite.jds

import com.javaworld.NamedPreparedStatement
import io.github.subiyacryolite.jds.enums.JdsImplementation
import java.sql.Connection
import java.util.*

/**
 * Created by ifunga on 14/07/2017.
 */
abstract class JdsDbOracle : JdsDb(JdsImplementation.ORACLE, true) {

    override fun tableExists(connection: Connection, tableName: String): Int {
        var toReturn = 0
        val sql = "SELECT COUNT(*) AS Result FROM all_objects WHERE object_type IN ('TABLE') AND object_name = ?"
        try {
            connection.prepareStatement(sql).use {
                it.setString(1, tableName.toUpperCase())
                it.executeQuery().use {
                    while (it.next())
                        toReturn = it.getInt("Result")
                }
            }
        } catch (ex: Exception) {
            toReturn = 0
            ex.printStackTrace(System.err)
        }
        return toReturn
    }

    override fun viewExists(connection: Connection, viewName: String): Int {
        var toReturn = 0
        val sql = "SELECT COUNT(*) AS Result FROM all_objects WHERE object_type IN ('VIEW') AND object_name = ?"
        try {
            connection.prepareStatement(sql).use {
                it.setString(1, viewName.toUpperCase())
                it.executeQuery().use {
                    while (it.next())
                        toReturn = it.getInt("Result")
                }
            }
        } catch (ex: Exception) {
            toReturn = 0
            ex.printStackTrace(System.err)
        }
        return toReturn
    }

    override fun procedureExists(connection: Connection, procedureName: String): Int {
        var toReturn = 0
        val sql = "SELECT COUNT(*) AS Result FROM all_objects WHERE object_type IN ('PROCEDURE') AND object_name = ?"
        try {
            connection.prepareStatement(sql).use {
                it.setString(1, procedureName.toUpperCase())
                it.executeQuery().use {
                    while (it.next())
                        toReturn = it.getInt("Result")
                }
            }
        } catch (ex: Exception) {
            toReturn = 0
            ex.printStackTrace(System.err)
        }
        return toReturn
    }

    override fun columnExists(connection: Connection, tableName: String, columnName: String): Int {
        var toReturn = 0
        val sql = "SELECT COUNT(COLUMN_NAME) AS Result FROM ALL_TAB_COLUMNS WHERE TABLE_NAME = :tableName AND COLUMN_NAME = :columnName"
        try {
            NamedPreparedStatement(connection, sql).use {
                it.setString("tableName", tableName.toUpperCase())
                it.setString("columnName", columnName.toUpperCase())
                it.executeQuery().use {
                    while (it.next())
                        toReturn = it.getInt("Result")
                }
            }
        } catch (ex: Exception) {
            toReturn = 0
            ex.printStackTrace(System.err)
        }
        return toReturn
    }

    override fun createStoreEntities(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_entity.sql")
    }

    override fun createRefEnumValues(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_enum.sql")
    }

    override fun createRefFields(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_field.sql")
    }

    override fun createRefFieldTypes(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_field_type.sql")
    }

    override fun createBindEntityFields(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_entity_field.sql")
    }

    override fun createBindEntityEnums(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_entity_enum.sql")
    }

    override fun createRefEntityOverview(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_entity_overview.sql")
    }

    override fun createBindEntityBinding(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_entity_binding.sql")
    }

    override fun createRefInheritance(connection: Connection) {
        executeSqlFromFile(connection, "sql/oracle/jds_ref_entity_inheritance.sql")
    }

    override fun getNativeDataTypeFloat(): String {
        return "BINARY_FLOAT"
    }

    override fun getNativeDataTypeDouble(): String {
        return "BINARY_DOUBLE"
    }

    override fun getNativeDataTypeZonedDateTime(): String {
        return "TIMESTAMP WITH TIME ZONE"
    }

    override fun getNativeDataTypeTime(): String {
        return "NUMBER(19)"
    }

    override fun getNativeDataTypeBlob(max: Int): String {
        return "BLOB"
    }

    override fun getNativeDataTypeInteger(): String {
        return "NUMBER(10)"
    }

    override fun getNativeDataTypeDate(): String {
        return "DATE"
    }

    override fun getNativeDataTypeDateTime(): String {
        return "TIMESTAMP"
    }

    override fun getNativeDataTypeLong(): String {
        return "NUMBER(19)"
    }

    override fun getNativeDataTypeString(max: Int): String {
        return if (max == 0) "NCLOB" else "NVARCHAR2($max)"
    }

    override fun getNativeDataTypeBoolean(): String {
        return "NUMBER(1,0)"
    }

    override fun getDbCreateIndexSyntax(tableName: String, columnName: String, indexName: String): String {
        return "CREATE INDEX $indexName ON $tableName($columnName);"
    }

    override fun createOrAlterProc(procedureName: String,
                                   tableName: String,
                                   columns: Map<String, String>,
                                   uniqueColumns: Collection<String>,
                                   doNothingOnConflict: Boolean): String {
        val sqlBuilder = StringBuilder()

        sqlBuilder.append("CREATE OR REPLACE PROCEDURE $procedureName(")
        val inputParameters = StringJoiner(", ")
        columns.forEach { column, type ->
            var shortType = type
            val containsBrackets = type.contains("(")
            if (containsBrackets)
                shortType = shortType.substring(0, shortType.indexOf('('))
            inputParameters.add("p_$column IN $shortType")
        }
        sqlBuilder.append(inputParameters)
        sqlBuilder.append(")\n")

        sqlBuilder.append("\t AS\n")
        sqlBuilder.append("\t\tBEGIN\n")
        sqlBuilder.append("\t\t\tMERGE  INTO $tableName dest\n")

        sqlBuilder.append("\t\t\tUSING DUAL ON (")
        val comparisonColumns = StringJoiner(" AND ")
        uniqueColumns.forEach { column -> comparisonColumns.add("$column = p_$column") }
        sqlBuilder.append(comparisonColumns)
        sqlBuilder.append(")\n")

        sqlBuilder.append("\t\t\tWHEN NOT MATCHED THEN INSERT (")
        val notMatchedColumns = StringJoiner(", ")
        columns.forEach { column, _ -> notMatchedColumns.add(column) }
        sqlBuilder.append(notMatchedColumns)
        sqlBuilder.append(")")

        sqlBuilder.append(" VALUES(")
        val notMatchedColumnsSrc = StringJoiner(", ")
        columns.forEach { column, _ -> notMatchedColumnsSrc.add("p_$column") }
        sqlBuilder.append(notMatchedColumnsSrc)
        sqlBuilder.append(")\n")

        if (!doNothingOnConflict && columns.count() > uniqueColumns.count()) {
            sqlBuilder.append("\t\t\tWHEN MATCHED THEN UPDATE SET ")
            val comparisonColumns = StringJoiner(", ")
            columns.forEach { column, _ ->
                if (!uniqueColumns.contains(column)) //dont update unique columns
                    comparisonColumns.add("$column = p_$column")
            }
            sqlBuilder.append(comparisonColumns)
        }
        sqlBuilder.append("\t\t\t;\n")
        sqlBuilder.append("\t\tEND $procedureName;")
        return sqlBuilder.toString()
    }
}
